packages/melos/lib/src/workspace.dart

```diff
@@ -48,15 +48,12 @@ class MelosWorkspace {
     required this.logger,
   });
 
-  /// Build a [MelosWorkspace] from a Directory.
-  /// If the directory is not a valid Melos workspace (e.g. no "melos.yaml" file)
-  /// then null is returned.
-  static Future<MelosWorkspace> fromDirectory(
```

Comment:
I would keep both constructors. I don't think there's a real value in changing existing usage of `fromDirectory` to `fromConfig`.

We could instead of `fromDirectory` use `fromConfig`

Reply:
For #163, the config has to be read early on before creating a workspace or even before parsing command-line arguments. Even though `MelosWorkspace.fromDirectory` is a convenient way to construct a workspace, calling it would result in parsing the config again.

Reply:
Yes I agree. But #163 is an edge-case. For more common cases, `fromDirectory` is fine, and switching to `fromConfig` is unnecessary

Reply:
The next step would be to read the config once at the top and pass it down to the command runner: https://github.com/jpnurmi/melos/commit/c98f9fbc8e76ee262c791a67bfed5b35d0d99d38

Since a workspace config provides the workspace path, there's no need to assume `Directory.current` down the tree anymore. Does this make it sound more reasonable to always instantiate workspaces from that config?

The last step (#163) would then look something like this: https://github.com/jpnurmi/melos/commit/8c6a5a55671d61f1fdfb514d8b73670628e12c86

Reply:
By the way, if you, for any reason, would prefer to solve #163 yourself, I'm happy to hand it over to you. :)

Reply:
Thanks for the reply!

> Since a workspace config provides the workspace path, there's no need to assume Directory.current down the tree anymore.

I'm not sure that this is a good thing. 
I could see commands have an argument for changing the working directory / specifying the path to melos.yaml, in which case we'd have to revert the change.

Similarly, MelosWorkspace.fromDirectory is a public API. Removing it is a breaking change, as people can call Melos commands programatically instead of relying on the CLI

Reply:
> I could see commands have an argument for changing the working directory / specifying the path to melos.yaml, in which case we'd have to revert the change.

Sounds even better reason to keep track of the top-level config path.

> Similarly, MelosWorkspace.fromDirectory is a public API. Removing it is a breaking change, as people can call Melos commands programatically instead of relying on the CLI

Aren't you about to release a new major version?

Reply:
> > Similarly, MelosWorkspace.fromDirectory is a public API. Removing it is a breaking change, as people can call Melos commands programatically instead of relying on the CLI
> 
> Aren't you about to release a new major version?

This API isn't public yet on a non-dev release yet since we're introducing it as part of v1 - so I don't think this would be a breaking change yet 🤔

Reply:
Oh, my bad I thought we already had a stable 1.0 release. You can ignore me then.